server/presenters/unfurl.ts

```diff
@@ -0,0 +1,209 @@
+import { differenceInMinutes, formatDistanceToNowStrict } from "date-fns";
+import { t } from "i18next";
+import { UnfurlResourceType, UnfurlResponse } from "@shared/types";
+import { dateLocale } from "@shared/utils/date";
+import { Document, User, View } from "@server/models";
+import { opts } from "@server/utils/i18n";
+import { GitHubUtils } from "plugins/github/shared/GitHubUtils";
+
+async function presentUnfurl(data: Record<string, any>) {
+  switch (data.type) {
+    case UnfurlResourceType.Mention:
+      return presentMention(data);
+    case UnfurlResourceType.Document:
+      return presentDocument(data);
+    case UnfurlResourceType.PR:
+      return presentPR(data);
+    case UnfurlResourceType.Issue:
+      return presentIssue(data);
+    default:
+      return presentOEmbed(data);
+  }
+}
+
+const presentOEmbed = (
+  data: Record<string, any>
+): UnfurlResponse[UnfurlResourceType.OEmbed] => ({
+  type: UnfurlResourceType.OEmbed,
+  url: data.url,
+  title: data.title,
+  description: data.description,
+  thumbnailUrl: data.thumbnail_url,
+});
+
+const presentMention = async (
+  data: Record<string, any>
+): Promise<UnfurlResponse[UnfurlResourceType.Mention]> => {
+  const user: User = data.user;
+  const document: Document = data.document;
+
+  const lastOnlineInfo = presentLastOnlineInfoFor(user);
+  const lastViewedInfo = await presentLastViewedInfoFor(user, document);
+
+  return {
+    type: UnfurlResourceType.Mention,
+    name: user.name,
+    avatarUrl: user.avatarUrl,
+    color: user.color,
+    lastActive: `${lastOnlineInfo} • ${lastViewedInfo}`,
+  };
+};
+
+const presentDocument = (
+  data: Record<string, any>
+): UnfurlResponse[UnfurlResourceType.Document] => {
+  const document: Document = data.document;
+  const viewer: User = data.viewer;
+  return {
+    url: document.url,
+    type: UnfurlResourceType.Document,
+    id: document.id,
+    title: document.titleWithDefault,
+    summary: document.getSummary(),
+    lastActivityByViewer: presentLastActivityInfoFor(document, viewer),
+  };
+};
+
+const presentPR = (
+  data: Record<string, any>
+): UnfurlResponse[UnfurlResourceType.PR] => ({
+  url: data.html_url,
+  type: UnfurlResourceType.PR,
+  id: `#${data.number}`,
+  title: data.title,
+  description: data.body,
+  author: {
+    name: data.user.login,
+    avatarUrl: data.user.avatar_url,
+  },
+  state: {
+    name: data.merged ? "merged" : data.state,
+    color: GitHubUtils.getColorForStatus(data.merged ? "merged" : data.state),
+  },
+  createdAt: data.created_at,
+});
+
+const presentIssue = (
+  data: Record<string, any>
+): UnfurlResponse[UnfurlResourceType.Issue] => ({
+  url: data.html_url,
+  type: UnfurlResourceType.Issue,
+  id: `#${data.number}`,
+  title: data.title,
+  description: data.body_text,
+  author: {
+    name: data.user.login,
+    avatarUrl: data.user.avatar_url,
+  },
+  labels: data.labels.map((label: { name: string; color: string }) => ({
```

Comment:
While overall I like the changes in this PR it's disappointing that the "custom" metadata fields are now all mixed in with the standard ones. I did like the clarity of separation between those fields that are expected on all responses

Reply:
Rationale here was to eliminate the barrier of decision-making - what should go inside `meta` and what not. Should `author` be inside `meta`? What about `createdAt`?

To be honest, `meta` just seemed to be a "hack" attribute to dump in resource specific attributes. Now that we've come with different types altogether, it perhaps eliminates the need for it? We still have the flexibility to put it for upcoming integrations if we've good reasons.